application-forum-ui/src/main/resources/ForumCode/RecentForumDiscussionsMacro.xml

```diff
@@ -0,0 +1,418 @@
+      </contentType>
+      <defaultCategory>
+        <disabled>0</disabled>
+        <name>defaultCategory</name>
+        <number>4</number>
+        <prettyName>Default category</prettyName>
+        <size>30</size>
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.StringClass</classType>
+      </defaultCategory>
+      <description>
+        <disabled>0</disabled>
+        <name>description</name>
+        <number>3</number>
+        <prettyName>Macro description</prettyName>
+        <rows>5</rows>
+        <size>40</size>
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.TextAreaClass</classType>
+      </description>
+      <id>
+        <disabled>0</disabled>
+        <name>id</name>
+        <number>1</number>
+        <prettyName>Macro id</prettyName>
+        <size>30</size>
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.StringClass</classType>
+      </id>
+      <name>
+        <disabled>0</disabled>
+        <name>name</name>
+        <number>2</number>
+        <prettyName>Macro name</prettyName>
+        <size>30</size>
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.StringClass</classType>
+      </name>
+      <supportsInlineMode>
+        <disabled>0</disabled>
+        <displayFormType>select</displayFormType>
+        <displayType>yesno</displayType>
+        <name>supportsInlineMode</name>
+        <number>5</number>
+        <prettyName>Supports inline mode</prettyName>
+        <unmodifiable>0</unmodifiable>
+        <classType>com.xpn.xwiki.objects.classes.BooleanClass</classType>
+      </supportsInlineMode>
+      <visibility>
+        <cache>0</cache>
+        <disabled>0</disabled>
+        <displayType>select</displayType>
+        <multiSelect>0</multiSelect>
+        <name>visibility</name>
+        <number>6</number>
+        <prettyName>Macro visibility</prettyName>
+        <relationalStorage>0</relationalStorage>
+        <separator>|</separator>
+        <separators>|</separators>
+        <size>1</size>
+        <unmodifiable>0</unmodifiable>
+        <values>Current User|Current Wiki|Global</values>
+        <classType>com.xpn.xwiki.objects.classes.StaticListClass</classType>
+      </visibility>
+    </class>
+    <property>
+      <code>{{velocity}}
+#set ($discard = $xwiki.ssx.use('ForumCode.RecentForumDiscussionsMacro'))
+#set ($limit = $xcontext.macro.params.limit)
+#if (!$limit)
+  #set ($limit = 5)
+#end
+## get from all the forums most recent N topics and topics' answers
+#set ($query = ", BaseObject as topic, BaseObject as answer where doc.fullName in (topic.name, answer.name) and doc.fullName not in ('ForumCode.TopicTemplate', 'ForumCode.AnswerTemplate') and topic.className='ForumCode.TopicClass' and answer.className='ForumCode.AnswerClass' order by doc.date desc")
```

Comment:
Also, I don't think you need 2 from clauses (1 for topic and one for answer) and you could simply use a generic `, BaseObject AS topicOrAnswer WHERE ... AND topicOrAnswer.className IN ('ForumCode.TopicClass', 'ForumCode.AnswerClass') ...`.

Note that this would also cover the above comment (since the IN clause implies an OR operation).